doc/design/GeneratedClientDesign.md

```diff
@@ -0,0 +1,133 @@
+Optional Property | get/set | get-only | get/set | [autorest.csharp #2339](https://github.com/Azure/autorest.csharp/issues/2339)
+Collection Property | `IList<T>` get-only | `IReadOnlyList<T>` get-only | `IList<T>` get-only | [autorest.csharp #2471](https://github.com/Azure/autorest.csharp/issues/2471)
+
+### Model Constructors
+
+Models have one or more constructors as follows.
+
+#### Main Constructor
+
+- Accessibility is specified in model shape table above
+- Takes required parameters and does not take optional parameters
+- Validates required reference type parameters for null using `Argument.AssertNotNull`
+- Takes list properties as `IEnumerable<T>` parameters
+- Initializes lists from `IEnumerable` parameter using System.Linq `.ToList()`
+
+#### Serialization Constructor
+
+- Internal accessibility
+- Generated for Output and Round-trip types
+- Takes required and optional parameters
+- Takes list properties as `IReadOnlyList<T>` parameters for Output models and `IList<T>` parameters for Round-trip models
+- List properties are initialized by assignment
+
+See examples:
+
+- Reference and value type properties:
+  - [RoundTripModel.cs](https://github.com/annelo-msft/azure-sdk-for-net/blob/cadl-models-roundtrip-basic/sdk/template/Azure.Template/src/Generated/Models/RoundTripModel.cs)
+  - [RoundTripModel.Serialization.cs](https://github.com/annelo-msft/azure-sdk-for-net/blob/cadl-models-roundtrip-basic/sdk/template/Azure.Template/src/Generated/Models/RoundTripModel.Serialization.cs)
+- Collection properties:
+  - [RoundTripModel.cs](https://github.com/annelo-msft/azure-sdk-for-net/blob/cadl-models-collections-basic/sdk/template/Azure.Template/src/Generated/Models/RoundTripModel.cs)
+  - Collection properties [RoundTripModel.Serialization.cs](https://github.com/annelo-msft/azure-sdk-for-net/blob/cadl-models-collections-basic/sdk/template/Azure.Template/src/Generated/Models/RoundTripModel.Serialization.cs)
+
+### Enums
+
+Enums may be generated as a CLR enum or a C# "Strongly-typed string" as follows.
+
+- Enums are generated in the `.Models` namespace
+- A Cadl `enum` not found in a corresponding  `@knownValues` decorator is generated as a C# `enum`
+- Generated enum has corresponding internal extensions file providing conversions to/from string.
+- A Cadl string model with a `@knownValues` decorator is generated as a C# "Strongly-typed string" (sometimes called "extensible enum") and the corresponding Cadl `enum` is not generated as a C# `enum`
+
+GitHub issue: [autorest.csharp #2477](https://github.com/Azure/autorest.csharp/issues/2477)
+
+See examples:
+
+- CLR enum:
+  - [DayOfTheWeek.cs](https://github.com/annelo-msft/azure-sdk-for-net/blob/cadl-models-enum-properties/sdk/template/Azure.Template/src/Generated/Models/DayOfTheWeek.cs)
+  - [DayOfTheWeek.Serialization.cs](https://github.com/annelo-msft/azure-sdk-for-net/blob/cadl-models-enum-properties/sdk/template/Azure.Template/src/Generated/Models/DayOfTheWeek.Serialization.cs)
```

Comment:
Add a deserialization methoid that takes what Json Dom returns.

Reply:
See https://github.com/Azure/autorest.csharp/issues/2477#issuecomment-1198696539